rust/mlt-nom/src/main.rs

```diff
@@ -0,0 +1,503 @@
+use nom::Err::Error as NomError;
+use nom::bytes::complete::take;
+use nom::combinator::complete;
+use nom::error::{Error, ErrorKind};
+use nom::multi::many0;
+use nom::{IResult, Parser};
+use num_enum::TryFromPrimitive;
+
+mod parsers;
+
+/// A layer that can be either MVT-compatible or unknown
+#[derive(Debug, PartialEq)]
+pub enum Layer<'a> {
+    /// MVT-compatible layer (tag = 1)
+    LayerV1(LayerV1<'a>),
+    /// Unknown layer with tag, size, and value
+    Unknown(Unknown<'a>),
+}
+
+impl Layer<'_> {
+    /// Parse a single binary tuple: size (varint), tag (varint), value (bytes)
+    pub fn parse(input: &[u8]) -> IResult<&[u8], Layer<'_>> {
+        let (input, size) = parsers::parse_varint_usize(input)?;
+
+        // tag is a varint, but we know fewer than 127 tags for now,
+        // so we can use a faster u8 and fail if it is bigger than 127.
+        let (input, tag) = parsers::parse_u8(input)?;
+        // 1 byte must be parsed for the tag, so if size is 0, it's invalid
+        let size = size
+            .checked_sub(1)
+            .ok_or(NomError(Error::new(input, ErrorKind::Fail)))?;
+        let (input, value) = take(size)(input)?;
+
+        let layer = match tag {
+            1 => {
+                let (data, meta) = LayerMeta::parse(value)?;
+                LayerV1::parse(data, meta)?
+            }
+            tag => Layer::Unknown(Unknown { tag, value }),
+        };
+
+        Ok((input, layer))
+    }
+}
+
+/// MVT-compatible layer data
+#[derive(Debug, PartialEq)]
+pub struct LayerV1<'a> {
+    pub meta: LayerMeta<'a>,
+    pub data: &'a [u8],
+}
+
+impl LayerV1<'_> {
+    #[expect(clippy::unnecessary_wraps)]
+    pub fn parse<'a>(
+        input: &'a [u8],
+        meta: LayerMeta<'a>,
+    ) -> Result<Layer<'a>, nom::Err<Error<&'a [u8]>>> {
+        for column in &meta.columns {
+            #[expect(clippy::match_same_arms)]
+            match column.typ {
+                ColumnType::Id => {
+                    // TODO: parse id
+                }
+                ColumnType::Geometry => {
+                    // TODO
+                }
+                ColumnType::StringProperty => {
+                    // TODO
+                }
+                ColumnType::FloatProperty => {
+                    // TODO
+                }
+                ColumnType::DoubleProperty => {
+                    // TODO
+                }
+                ColumnType::IntProperty => {
+                    // TODO
+                }
+                ColumnType::UintProperty => {
+                    // TODO
+                }
+                ColumnType::SintProperty => {
+                    // TODO
+                }
+                ColumnType::BoolProperty => {
+                    // TODO
```

Comment:
Is this what we want?

```suggestion
                    todo!()
                }
                ColumnType::Geometry => {
                    todo!()
                }
                ColumnType::StringProperty => {
                    todo!()
                }
                ColumnType::FloatProperty => {
                    todo!()
                }
                ColumnType::DoubleProperty => {
                    todo!()
                }
                ColumnType::IntProperty => {
                    todo!()
                }
                ColumnType::UintProperty => {
                    todo!()
                }
                ColumnType::SintProperty => {
                    todo!()
                }
                ColumnType::BoolProperty => {
                    todo!()
```